### ISSUE TITLE
Http Code 100 problem with htx

### ISSUE THREAD

**Issue by michalkalisz** [haproxy/haproxy#27]
## Output of `haproxy -vv` and `uname -a`

```
$ docker exec lb-test haproxy -vv
HA-Proxy version 1.9.2 2019/01/16 - https://haproxy.org/
Build options :
  TARGET  = linux2628
  CPU     = generic
  CC      = gcc
  CFLAGS  = -O2 -g -fno-strict-aliasing -Wdeclaration-after-statement -fwrapv -Wno-unused-label -Wno-sign-compare -Wno-unused-parameter -Wno-old-style-declaration -Wno-ignored-qualifiers -Wno-clobbered -Wno-missing-field-initializers -Wtype-limits -Wshift-negative-value -Wshift-overflow=2 -Wduplicated-cond -Wnull-dereference
  OPTIONS = USE_ZLIB=1 USE_OPENSSL=1 USE_LUA=1 USE_PCRE=1

Default settings :
  maxconn = 2000, bufsize = 16384, maxrewrite = 1024, maxpollevents = 200

Built with OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
Running on OpenSSL version : OpenSSL 1.1.0j  20 Nov 2018
OpenSSL library supports TLS extensions : yes
OpenSSL library supports SNI : yes
OpenSSL library supports : TLSv1.0 TLSv1.1 TLSv1.2
Built with Lua version : Lua 5.3.3
Built with transparent proxy support using: IP_TRANSPARENT IPV6_TRANSPARENT IP_FREEBIND
Built with zlib version : 1.2.8
Running on zlib version : 1.2.8
Compression algorithms supported : identity("identity"), deflate("deflate"), raw-deflate("deflate"), gzip("gzip")
Built with PCRE version : 8.39 2016-06-14
Running on PCRE version : 8.39 2016-06-14
PCRE library supports JIT : no (USE_PCRE_JIT not set)
Encrypted password support via crypt(3): yes
Built with multi-threading support.

Available polling systems :
      epoll : pref=300,  test result OK
       poll : pref=200,  test result OK
     select : pref=150,  test result OK
Total: 3 (3 usable), will use epoll.

Available multiplexer protocols :
(protocols marked as <default> cannot be specified using 'proto' keyword)
              h2 : mode=HTX        side=FE|BE
              h2 : mode=HTTP       side=FE
       <default> : mode=HTX        side=FE|BE
       <default> : mode=TCP|HTTP   side=FE|BE

Available filters :
	[SPOE] spoe
	[COMP] compression
	[CACHE] cache
	[TRACE] trace


```

## What's the configuration?

```
global
    stats timeout 30s
    daemon
    maxconn 65536

defaults
    log	global
    option http-use-htx
    option httplog
    option dontlognull
    option log-health-checks
    timeout connect 10s
    timeout client  60m
    timeout server  60m

    retries 2
    option redispatch
    maxconn 65536
    option http-keep-alive
    http-reuse always
    timeout client-fin 30s
    timeout http-keep-alive 10s


frontend test-1
    mode http
    option forwardfor

    log global
    bind *:80

    default_backend test-backend 

backend test-backend 
    mode http
    server echo_server postman-echo.com:80 check weight 100 maxconn 64 slowstart 1s inter 5000

listen  stats-1
    bind  *:7654
    mode http
    log global 
    stats enable
    stats uri /
    stats refresh 5s
    bind-process 1

```

## Steps to reproduce the behavior

send http request with Expect: 100-continue, i.e.:
```
curl -v -L --request POST "http://127.0.0.1/post" -d @file_with_more_than_1000characters
```
or
``` 
curl -v -L -H "Expect: 100-continue" --request POST "http://127.0.0.1/post" -d "Test"
```

To run haproxy in docker:

[code_100_test.zip](https://github.com/haproxy/haproxy/files/2813142/code_100_test.zip)
extract zip  
```
cd code_100_test
docker build -t haproxy .
docker run --rm --network host --name lb-test haproxy
```
to compare with 1.8 change cwd to 1_8

## Actual behavior

Haproxy response 100 code with
transfer-encoding: chunked

Result from curl:
```
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> POST /post HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.47.0
> Accept: */*
> Expect: 100-continue
> Content-Length: 4
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 100 Continue
< transfer-encoding: chunked
* Illegal or missing hexadecimal sequence in chunked-encoding
* Closing connection 0
curl: (56) Illegal or missing hexadecimal sequence in chunked-encoding
```
Tcpdump:
```
15:34:32.215728 IP localhost.57088 > localhost.http: Flags [S]
15:34:32.215744 IP localhost.http > localhost.57088: Flags [S.]
15:34:32.215757 IP localhost.57088 > localhost.http: Flags [.]
15:34:32.215810 IP localhost.57088 > localhost.http: Flags [P.]
POST /post HTTP/1.1
Host: 127.0.0.1
User-Agent: curl/7.47.0
Accept: */*
Expect: 100-continue
Content-Length: 4
Content-Type: application/x-www-form-urlencoded


15:34:32.417391 IP localhost.http > localhost.57088: Flags [P.]HTTP/1.1 100 Continue
transfer-encoding: chunked

0


15:34:32.417405 IP localhost.57088 > localhost.http: Flags [.]
15:34:32.417545 IP localhost.57088 > localhost.http: Flags [F.]
15:34:32.417682 IP localhost.http > localhost.57088: Flags [P.]
HTTP/1.0 400 Bad request
cache-control: no-cache
content-type: text/html

<html><body><h1>400 Bad request</h1>
Your browser sent an invalid request.
</body></html>

15:34:32.417717 IP localhost.57088 > localhost.http: Flags [R]
```

## Expected behavior

1. curl send req
2. haproxy send req to server
3. haproxy recv response from server (100)
4. haproxy send resp to client (100)
5. curl send req body
6. haproxy send req body to server
7. haproxy recv response from server (200)
8. haproxy send response to client(200)

Result from curl (for haproxy 1.8)
```
*   Trying 127.0.0.1...
* Connected to 127.0.0.1 (127.0.0.1) port 80 (#0)
> POST /post HTTP/1.1
> Host: 127.0.0.1
> User-Agent: curl/7.47.0
> Accept: */*
> Expect: 100-continue
> Content-Length: 4
> Content-Type: application/x-www-form-urlencoded
> 
< HTTP/1.1 100 Continue
* We are completely uploaded and fine
< HTTP/1.1 200 OK
< Content-Type: application/json; charset=utf-8
< Date: Wed, 30 Jan 2019 14:47:47 GMT
< ETag: W/"12a-msUD0LWLNBNDs5uc/Ir6S5Zxlcs"
< Server: nginx
< set-cookie: sails.sid=s%3Ahfts6SdufXP0gZvu4Y0PM1aGjB2UhuHg.X9sH64LDXaDy0NjdSXgjDhrYJPWmu8KlSKMRytVp54M; Path=/; HttpOnly
< Vary: Accept-Encoding
< Content-Length: 298
< 
* Connection #0 to host 127.0.0.1 left intact
{"args":{},"data":"","files":{},"form":{"Test":""},"headers":{"x-forwarded-proto":"https","host":"127.0.0.1","content-length":"4","accept":"*/*","content-type":"application/x-www-form-urlencoded","user-agent":"curl/7.47.0","x-forwarded-port":"80"},"json":{"Test":""},"url":"https://127.0.0.1/post"}
```

**Comment by capflam** [comment]
I confirm the bug. It was introduced by the commit 4710d20743. I'm working on a fix. It will be pretty simple to do so. 
Thanks.

**Comment by wtarreau** [comment]
Hmmm that's me again, sorry guys!

**Comment by capflam** [comment]
In fact the bug did not affect only 100 responses. The header "transfer-encoding" header was also added in responses to HEAD/CONNECT requests and all 1xx, 204 and 304 responses. I pushed a fix in 2.0-dev. Thanks !

**Comment by capflam** [comment]
Reopen it. Must be backported in 1.9

**Comment by wtarreau** [comment]
now merged into 1.9.

**Comment by michalkalisz** [comment]
Thanks guys!

**Comment by steveum0105** [comment]
Hi guys. I see this Issue again on 1.9.8 with htx mode and HTTP/1.1.
HTTP Request is hanging after server response 'HTTP/1.1 100 Continue'.

Try this.

```
dd if=/dev/zero bs=1 count=1025 | curl -vvv --data-binary "@-" http://any.your.domain.com
```

Then I see

```
> POST / HTTP/1.1
> User-Agent: curl/7.29.0
> Host: ---
> Accept: */*
> Content-Length: 1025
> Content-Type: application/x-www-form-urlencoded
> Expect: 100-continue
>
< HTTP/1.1 100 Continue
...
```

**Comment by steveum0105** [comment]
When I try the same on 1.8.14, It works fine.

**Comment by steveum0105** [comment]
When I try this without option htx, It works fine.

**Comment by capflam** [comment]
@steveum0105, I pushed 2 fixes into the haproxy-1.9 repository :

  * http://git.haproxy.org/?p=haproxy-1.9.git;a=commit;h=6d30abbb
  * http://git.haproxy.org/?p=haproxy-1.9.git;a=commit;h=674e755d

It should be ok now. Thanks !
